canonical-web-and-design/maas-ui
# Spike mock server devtools 1083

## Done

- proof of concept - websocket servering mock data (read-only)
- add `MAAS_DEVTOOLS` that allow to manually send websocket messages from the browser console and log data useful for debugging
- allows to switch between different scenarios for quick preview of a UI (using the same websocket messages that the real back-end sends)
- this mock server would be sufficient in many cases when doing QA
- it would also allow to run maas-ui without having to run resource-hungry maas back-end and work offline without being able to connect to our remote instances (e.g. bolla).

Note: This is is a spike, treat this as a proof of concept - the implementation is somewhat dirty.

## QA

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Karura](/HACKING.md#karura)
- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

### QA steps
- Go to the deployed demo (link in the comments), or run locally:
  - set `MAAS_URL=http://localhost:8080/` in your `.env.local`
  - run `yarn start-mock-server-and-serve`
- login: `admin` password: `test1`
- run `MAAS_DEVTOOLS.enable()`
- refresh the page
- try sending an (incoming) websocket message and see the UI update accordingly, e.g. `MAAS_DEVTOOLS.send("machine.list")`
- use the dropdown in the status bar to switch between scenarios

## Fixes

Fixes:

## Launchpad issue

Related Launchpad maas issue in the form `lp#number`.

## Backports

In general, please propose fixes against _main_ rather than release branches (e.g. 2.7), unless the fix is only applicable for that specific release. Please apply backport labels to the PR (e.g. "Backport 2.7") for the appropriate releases to target.

Only bug and security fixes should be backported, new features should only land in main.

## Screenshots

<img width="1032" alt="image" src="https://user-images.githubusercontent.com/7452681/176715701-49071dea-3db5-454f-b148-3e7cce25faec.png">
<img width="678" alt="image" src="https://user-images.githubusercontent.com/7452681/176717549-d4abe4c0-7e7a-4c25-8f85-92b0d84451b7.png">
